dbms/src/Flash/Mpp/MPPTunnel.cpp

```diff
@@ -106,67 +91,73 @@ MPPTunnelBase<Writer>::~MPPTunnelBase()
     LOG_FMT_TRACE(log, "destructed tunnel obj!");
 }
 
-template <typename Writer>
-void MPPTunnelBase<Writer>::finishSendQueue()
+void MPPTunnel::finishSendQueue()
 {
-    bool flag = send_queue.finish();
-    if (flag && !is_local && is_async)
-        writer->tryFlushOne();
+    bool flag = send_queue->finish();
+    if (flag && mode == TunnelSenderMode::ASYNC_GRPC)
+    {
+        async_tunnel_sender->tryFlushOne();
+    }
 }
 
 /// exit abnormally, such as being cancelled.
-template <typename Writer>
-void MPPTunnelBase<Writer>::close(const String & reason)
+void MPPTunnel::close(const String & reason)
 {
     {
         std::unique_lock lk(mu);
-        if (finished)
+        switch (status)
+        {
+        case TunnelStatus::Unconnected:
+            status = TunnelStatus::Finished;
+            cv_for_status_changed.notify_all();
             return;
-        if (connected)
+        case TunnelStatus::Connected:
+        case TunnelStatus::WaitingForSenderFinish:
         {
             if (!reason.empty())
             {
                 try
                 {
                     FAIL_POINT_TRIGGER_EXCEPTION(FailPoints::exception_during_mpp_close_tunnel);
-                    send_queue.push(std::make_shared<mpp::MPPDataPacket>(getPacketWithError(reason)));
-                    if (!is_local && is_async)
-                        writer->tryFlushOne();
+                    send_queue->push(std::make_shared<mpp::MPPDataPacket>(getPacketWithError(reason)));
```

Comment:
When in `WaitingForSenderFinish` state, I think the `send_queue` is already `finished` or `cancelled`?

Reply:
Seems in WaitingForSenderFinish state, MPPTunnel should just ignore the close event.